### ISSUE TITLE
Document that Ansible API is not thread-safe

### ISSUE THREAD

**Issue by jeblair** [ansible/ansible#13822]
When ansible loads a module from disk for a plugin, it first checks
to see if the module has already been loaded (i.e., exists in
sys.modules), and if it does, it returns that reference rather than
loading the module again (which, as described in issue #13110, must
not happen because repeating the module load may re-initialize
module global variables).    

When a module is loaded, the reference is immediately added to  
sys.modules, even before the module initialization is complete. 
Therefore, if two threads race the module loading code, the first
will begin loading the module from disk, and the second will see  
the module in sys.modules.  It might begin using that module before
it is even fully loaded, resulting in AttributeErrors on objects 
which should be defined in the module.

This can be reproduced by wrapping the Python API example at http://docs.ansible.com/ansible/developing_api.html in a function and then adding the following:

``` python
import threading
t1 = threading.Thread(target=test)
t1.start()
t2 = threading.Thread(target=test)
t2.start()
t1.join()
t2.join()
```

This results in the this traceback:

> Traceback (most recent call last):
>   File "/usr/lib/python2.7/threading.py", line 810, in **bootstrap_inner
>     self.run()
>   File "/usr/lib/python2.7/threading.py", line 763, in run
>     self.__target(_self.__args, *_self.__kwargs)
>   File "test.py", line 13, in test
>     variable_manager = VariableManager()
>   File "lib/ansible/vars/__init**.py", line 92, in **init**
>     self._fact_cache = FactCache()
>   File "lib/ansible/plugins/cache/__init__.py", line 35, in **init**
>     self._plugin = cache_loader.get(C.CACHE_PLUGIN)
>   File "lib/ansible/plugins/__init__.py", line 335, in get
>     obj = getattr(self._module_cache[path], self.class_name)(_args, *_kwargs)
> AttributeError: 'module' object has no attribute 'CacheModule'

The problem can be corrected by wrapping module loading in a mutex.

**Comment by jimi-c** [comment]
Updated title to reflect the conversation in #13823

**Comment by bcoca** [comment]
fixed via 2374abbeda465c0aded55fabc18085a1530c9377

**Comment by crazychengmm** [comment]
hi it‘s also happened  version 2.4.3.0,is it fixed?
@bcoca  @jeblair 
--------------
ansible 2.4.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.9 (default, Jun 29 2016, 13:08:31) [GCC 4.9.2]
----------------
I use the api do some commands on some hosts,but commands costs times,so i use threading.
command was runed on same host. at last get result from all thread.
some times that's worked,but some times faild,perhaps 50% to 50%.
----------------
here is faild error,and it has multiple error,not faild with same error.
A:
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/home/hzchengmiao/ansibleapitest.py", line 24, in run
    self.result = self.func(*self.args)
  File "/home/hzchengmiao/ansibleapitest.py", line 81, in exec_ansible
    inventory = InventoryManager(loader=loader, sources=['/etc/ansible/hosts'])
  File "/usr/local/lib/python2.7/dist-packages/ansible/inventory/manager.py", line 145, in __init__
    self.parse_sources()
  File "/usr/local/lib/python2.7/dist-packages/ansible/inventory/manager.py", line 206, in parse_sources
    parse = self.parse_source(source, cache=cache)
  File "/usr/local/lib/python2.7/dist-packages/ansible/inventory/manager.py", line 258, in parse_source
    if plugin.verify_file(source):
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/inventory/yaml.py", line 74, in verify_file
    if super(InventoryModule, self).verify_file(path):
TypeError: super(type, obj): obj must be an instance or subtype of type

B:
 [WARNING]: Unable to load the facts cache plugin (memory).
[WARNING]: Skipping plugin (/usr/local/lib/python2.7/dist-packages/ansible/plugins/callback/foreman.py) as it seems to be invalid: 'module' object has no attribute 'CallbackModule'

C:
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/home/hzchengmiao/ansibleapitest.py", line 24, in run
    self.result = self.func(*self.args)
  File "/home/hzchengmiao/ansibleapitest.py", line 81, in exec_ansible
    inventory = InventoryManager(loader=loader, sources=['/etc/ansible/hosts'])
  File "/usr/local/lib/python2.7/dist-packages/ansible/inventory/manager.py", line 145, in __init__
    self.parse_sources()
  File "/usr/local/lib/python2.7/dist-packages/ansible/inventory/manager.py", line 206, in parse_sources
    parse = self.parse_source(source, cache=cache)
  File "/usr/local/lib/python2.7/dist-packages/ansible/inventory/manager.py", line 285, in parse_source
    display.vvv(to_text(fail['exc'].tb))
AttributeError: 'exceptions.TypeError' object has no attribute 'tb' 

Thanks.
